src/ch/ch_monitor.c

```diff
@@ -563,25 +564,338 @@ chMonitorBuildSocketCmd(virDomainObjPtr vm, const char *socket_path)
+        return -1;
+    }
+    if (virJSONValueObjectHasKey(eventJSON, "event") == 0) {
+        VIR_WARN("Invalid JSON from monitor, no event key");
+        return -1;
+    }
+    source = virJSONValueObjectGetString(eventJSON, "source");
+    event = virJSONValueObjectGetString(eventJSON, "event");
+
+    ev = virCHMonitorEventFromString(source, event);
+    VIR_DEBUG("Source: %s Event: %s, ev: %d", source, event, ev);
+    switch (ev) {
+        case virCHMonitorVmEventBooted:
+        case virCHMonitorVmEventResumed:
+        case virCHMonitorVmEventRestored:
+        case virCHMonitorVirtioDeviceEventActivated:
+        case virCHMonitorVirtioDeviceEventReset:
+            virObjectLock(vm);
+            if (virDomainObjIsActive(vm) &&
+                   virCHDomainObjBeginJob(vm, CH_JOB_MODIFY) == 0) {
+                virCHProcessSetupThreads(vm);
+                virCHDomainObjEndJob(vm);
+            }
+            virObjectUnlock(vm);
+            break;
+        case virCHMonitorVmmEventShutdown: // shutdown inside vmm
+        case virCHMonitorVmEventShutdown:
+            {
+                virCHDriverPtr driver = CH_DOMAIN_PRIVATE(vm)->driver;
+                virDomainState state;
+
+                virObjectLock(vm);
+                state = virDomainObjGetState(vm, NULL);
+                if ((ev == virCHMonitorVmmEventShutdown ||
+                     state == VIR_DOMAIN_SHUTDOWN) &&
+                      (virCHDomainObjBeginJob(vm, CH_JOB_MODIFY) == 0)) {
+
+                    virCHProcessStop(driver, vm, VIR_DOMAIN_SHUTOFF_SHUTDOWN);
+
+                    virCHDomainObjEndJob(vm);
+                }
+                virObjectUnlock(vm);
+                break;
+            }
+        case virCHMonitorVmEventBooting:
+        case virCHMonitorVmEventPausing:
+        case virCHMonitorVmEventPaused:
+        case virCHMonitorVmEventResuming:
+        case virCHMonitorVmEventSnapshotting:
+        case virCHMonitorVmEventSnapshotted:
+        case virCHMonitorVmEventRestoring:
+        case virCHMonitorVmEventResizing:
+        case virCHMonitorVmEventResized:
+        case virCHMonitorVmEventDeleted:
+        case virCHMonitorVmmEventStarting:
+        case virCHMonitorCpuCreateVcpu:
+            break;
+        case virCHMonitorEventMax:
+        default:
+            VIR_WARN("unkown event from monitor!");
+            break;
+    }
+
+    return 0;
+}
+
+/*
+ * Helper function to find a block of valid JSON
+ * from a stream of multiple JSON blocks.
+ */
+static inline char *end_of_json(char *str, size_t len)
+{
+    bool started = false;
+    int blocks = 0;
+    int i = 0;
+
+    while ((i < len) && (!started || blocks > 0)) {
+        if (str[i] == '{') {
+            if (!started)
+                started = true;
+            blocks++;
+        } else if (str[i] == '}') {
+            blocks--;
+        }
+        i++;
+    }
+
+    return (i == len && blocks) ? NULL : str + i;
```

Comment:
Couldn't we have a full buffer that ends perfectly at a closing `}`. In that case this would return `str + i` as blocks would be 0 but `str + i` would be the location one past the end of the buffer wouldn't it would be an overflow (as we read/write to the that location in `virCHMonitorProcessEvents`)?

Reply:
yes you are right. Will fix!

Reply:
Hmm on going through the code again, I think we have a bigger problem. virJSONValueFromString requires that EOL('\0') to identify the end of the document. So we would need to grow the buffer to get it processed correctly. Thats not a scalable solution. So we would need to make sure that the JSON doc size should be atmost one less than the buf size. We shall discuss this during tomorrow's meeting and have the protocol guarentee that, the JSON doc is with in size limits. And we wiil consider a JSON doc invalid even if its more than he protocol size

Reply:
But there is another probability that we hit a smaller JSON doc which happens to be right at the end of the buffer. I will take care of this in the code.